code/model2.py

```diff
@@ -1,9 +1,129 @@
+        r_ids = autograd.Variable(torch.LongTensor(r_ids))
+        # extract the embedding for all the receivers and take an average
+        r_emb = torch.mean(self.embedding_layer(r_ids), 0, True)
+        # extract the sender embedding
+        s_emb = self.embedding_layer(s_id)
+        # print s_emb.size()
+        # print r_emb.size()
+        #print size(pv_emb)
+        p_emb = torch.autograd.Variable(torch.from_numpy(pv_emb))
+        p_emb = p_emb.unsqueeze(0)
+        n_emb = torch.autograd.Variable(torch.from_numpy(nv_emb))
+        n_emb = n_emb.unsqueeze(0)
+        # print n_emb.size()
+        # simple concatenation of sender and receiver embedding
+        # sr_emb = torch.cat((s_emb, r_emb), 1)
+        # pn_emb = torch.cat((p_emb,n_emb), 1)
+        # print sr_emb.size()
+        # print pn_emb.size()
+        srpn_emb = torch.cat([s_emb, r_emb,p_emb,n_emb], 1)
+        # srpn_emb = torch.cat((srp_emb, n_emb),1)
+        # print srpn_emb.size()
+        h1 = self.relu(self.h1_layer(srpn_emb))
+        word_reps = self.email_layer(h1)
+        return word_reps
+
+    def get_average_rep(self, word_reps):
+        """
+        Assumes that word_reps is a numpy 2d array with every row as vector representation of word.
+        Calculates the mean across all rows to get average email representation
+        :param word_reps:
+        :return:
+        """
+        return np.mean(word_reps, axis=0)
+
     def train(self, emails, w2v):
-        pass
+        loss_criteria = nn.MSELoss()
+        optimizer = optim.RMSprop(self.parameters(), lr=0.001, alpha=0.6, momentum=0.6)
+
+        for epoch in range(self.epochs):
+            epoch_loss = 0.0
+            for i in range(len(emails)):
+                sender_id = utils.get_userid(emails[i, 0])
+                email_content = emails[i, 2]
+                if sender_id is None or email_content is None:
+                    continue
+                recv_list = emails[i, 1].split('|')
+                recv_ids = []
+                for recv in recv_list:
+                    recv_id = utils.get_userid(recv)
+                    if recv_id is not None:
+                        recv_ids.append(recv_id)
+                # if none of the receivers were found, ignore this case
+                if len(recv_ids) == 0:
+                    continue
+                # print 'new mail'
+                # print '-------------'
+                # print email_content
+                #print '------------------\n\n'
+                #print i, email_content, sender_id, recv_ids
+                email_words = emails[i, 2].split()
+                for j in range(1,len(email_words)-1):
+					prev_word = email_words[j-1]
+					next_word = email_words[j+1]
+					email_word_rep = w2v.get_word(email_words[j]) 
+					pv_emb = w2v.get_word(prev_word)
+					nv_emb = w2v.get_word(next_word)
+					if pv_emb is None or nv_emb is None or email_word_rep is None:
```

Comment:
Instead of using get_word and then doing this, you can use get_sentence which returns embeddings for the entire sentence.
That way, you will drop very few combinations and end up using only valid word combinations.

Eg.

This is a sentence with an OUTOFVOCAB word in it
Trains only -> this is a, is a sentence, a sentence with, sentence with an, word in it (5 combos)

This is a sentence with an word in it (get sentence drops other words)
Trains -> (6 combos)

Reply:
but this could result in a problem. you only want to train using wi-1 and wi+1. If you directly use the output of get_scentence, you end up training using words that were originally not adjacent to each other in the dataset